src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -0,0 +1,532 @@
+  namespace packages and resource files in the source.
+  """
+  digest: Digest
+  packages: Tuple[str, ...]
+  namespace_packages: Tuple[str, ...]
+  package_data: Tuple[PackageDatum, ...]
+
+
+@dataclass(frozen=True)
+class SetupPyChrootRequest:
+  """A request to create a chroot containing a setup.py and the sources it operates on."""
+  exported_target: ExportedTarget
+  py2: bool = False  # Whether to use py2 or py3 namespace package semantics.
+
+
+@dataclass(frozen=True)
+class SetupPyChroot:
+  """A chroot containing a generated setup.py and the sources it operates on."""
+  digest: Digest
+  # The keywords are embedded in the setup.py file in the digest, so these aren't
+  # strictly needed here, but they are convenient for testing.
+  setup_keywords_json: str
+
+
+@dataclass(frozen=True)
+class RunSetupPyRequest:
+  """A request to run a setup.py command."""
+  exported_target: ExportedTarget
+  chroot: SetupPyChroot
+  args: Tuple[str, ...]
+
+
+@dataclass(frozen=True)
+class RunSetupPyResult:
+  """The result of running a setup.py command."""
+  output: Digest  # The state of the chroot after running setup.py.
+
+
+@dataclass(frozen=True)
+class SetuptoolsSetup:
+  """The setuptools tool."""
+  requirements_pex: Pex
+
+
+class SetupPyOptions(GoalSubsystem):
+  name = "setup-py2"
+
+  @classmethod
+  def register_options(cls, register):
+    super().register_options(register)
+    register(
+      '--args', type=list, member_type=str,
+      help="Arguments to pass directly to setup.py, e.g. "
+           "`--setup-py2-args=\"bdist_wheel\"`. If unspecified, we just dump the setup.py chroot."
+    )
+
+
+class SetupPy(Goal):
+  """Runs setup.py commands."""
+  subsystem_cls = SetupPyOptions
+
+
+@console_rule
+async def run_setup_pys(addresses: BuildFileAddresses, console: Console, options: SetupPyOptions,
+                        distdir: DistDir, workspace: Workspace) -> SetupPy:
+  """Run setup.py commands on all exported targets addressed."""
+  targets = await Get[HydratedTargets](BuildFileAddresses, addresses)
+
+  exported_targets: List[ExportedTarget] = []
+  nonexported_targets: List[HydratedTarget] = []
+  for target in targets:
+    if _is_exported(target):
+      exported_targets.append(ExportedTarget(target))
+    else:
+      nonexported_targets.append(target)
+  if nonexported_targets:
+    raise TargetNotExported(
+      'Cannot run setup.py on these targets, because they have no `provides=` clause: '
+      f'{", ".join(so.address.reference() for so in nonexported_targets)}')
+
+  for target in exported_targets:
+    addr = target.hydrated_target.address.reference()
+    chroot = await Get[SetupPyChroot](SetupPyChrootRequest(target))
+    provides = target.hydrated_target.adaptor.provides
+    setup_py_dir = distdir.relpath / f'{provides.name}-{provides.version}'
+    if options.values.args:
+      console.print_stderr(f'Running setup.py for {addr}')
```

Comment:
This might be a lie due to caching: would recommend moving it to a log statement inside of the rule that actually prepares the execution... which might _still_ be a lie due to the persistent cache, but which would at least avoid rendering it for every invoke under pantsd.

Reply:
I'll drop it entirely, for now anyway. The v2 ui will tell us when a process is actually running.